### PR TITLE
Add validate_args decorators to API functions in tmol.io

### PR DESCRIPTION
**Pull request by aleaverfay**
Add the `@validate_args` function to all API calls; add the most important classes to`__init__.py` 

If you import `tmol`, you should have access to `PoseStack` and `ScoreFunction` and the other classes you are most likely to pass around between functions in your external project.

This PR also adds type hints to all the API functions listed in tmol/tmol/__init__.py

### CODE REVIEW

**Comment by aleaverfay** [comment]
This PR is ready. @fdimaio @jflat06

File: tmol/io/canonical_ordering.py [hunk 1]
```diff
@@ -335,14 +383,19 @@ def _d(x):
         )
 
 
+@validate_args
 @toolz.functoolz.memoize
```
Comment (fdimaio) [inline tmol/io/canonical_ordering.py:387]: does validate_args do anything on functions w no arguments?  Does it validate return type?

File: tmol/tests/pack/rotamer/test_build_rotamers.py [hunk 2]
```diff
@@ -37,12 +37,12 @@ def test_annotate_restypes(
...
-        assert type(rt.rotamer_kinforest.parent) is numpy.ndarray
-        assert type(rt.rotamer_kinforest.frame_x) is numpy.ndarray
-        assert type(rt.rotamer_kinforest.frame_y) is numpy.ndarray
-        assert type(rt.rotamer_kinforest.frame_z) is numpy.ndarray
+        assert isinstance(rt.rotamer_kinforest.id, numpy.ndarray)
+        assert isinstance(rt.rotamer_kinforest.doftype, numpy.ndarray)
+        assert isinstance(rt.rotamer_kinforest.parent, numpy.ndarray)
+        assert isinstance(rt.rotamer_kinforest.frame_x, numpy.ndarray)
+        assert isinstance(rt.rotamer_kinforest.frame_y, numpy.ndarray)
+        assert isinstance(rt.rotamer_kinforest.frame_z, numpy.ndarray)
```
Comment (fdimaio) [inline tmol/tests/pack/rotamer/test_build_rotamers.py:45]: Is there an advantage to isinstance over type?

**Review by fdimaio — APPROVED** [review]
This looks great to me.

**Pull request merged**